src/Microsoft.Extensions.FileProviders.Physical/Watcher/DotnetFileWatcher.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+
+namespace Microsoft.Extensions.FileProviders.Physical.Watcher
+{
+    internal class DotnetFileWatcher : IFileSystemWatcher
+    {
+        private readonly Func<string, FileSystemWatcher> _watcherFactory;
+        private readonly string _watchedDirectory;
+
+        private FileSystemWatcher _fileSystemWatcher;
+
+        public DotnetFileWatcher(string watchedDirectory)
+            : this(watchedDirectory, DefaultWatcherFactory)
+        {
+        }
+
+        internal DotnetFileWatcher(string watchedDirectory, Func<string, FileSystemWatcher> fileSystemWatcherFactory)
+        {
+            if (string.IsNullOrEmpty(watchedDirectory))
+            {
+                throw new ArgumentNullException(nameof(watchedDirectory));
+            }
+
+            _watchedDirectory = watchedDirectory;
+            _watcherFactory = fileSystemWatcherFactory;
+            CreateFileSystemWatcher();
+        }
+
+        public event EventHandler<string> OnFileChange;
+
+        public event EventHandler OnError;
+
+        private static FileSystemWatcher DefaultWatcherFactory(string watchedDirectory)
+        {
+            if (string.IsNullOrEmpty(watchedDirectory))
+            {
+                throw new ArgumentNullException(nameof(watchedDirectory));
+            }
+
+            return new FileSystemWatcher(watchedDirectory);
+        }
+
+        private void FSW_Error(object sender, ErrorEventArgs e)
+        {
+            // Recreate the watcher
+            CreateFileSystemWatcher();
```

Comment:
Couldn't this end up recursing if the watcher raises `Error` all the time (say if network connection drops to a share that it was watching)?